File: omnexa_construction/omnexa_construction/page/primavera_project_import/primavera_project_import.js
frappe.pages["primavera-project-import"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Primavera Project Import"),
		single_column: true,
	});

	const state = { file_url: null, projects: [] };

	const $layout = $(`
		<div class="primavera-import-page">
			<p class="text-muted">${__(
				"Import projects from a Primavera P6 XER export file. Each project creates a Project Contract and Schedule Baseline."
			)}</p>
			<div class="row">
				<div class="col-md-4">
					<div class="form-group">
						<label>${__("Company")}</label>
						<div class="company-field"></div>
					</div>
				</div>
				<div class="col-md-4">
					<div class="form-group">
						<label>${__("Branch")}</label>
						<div class="branch-field"></div>
					</div>
				</div>
				<div class="col-md-4">
					<div class="form-group">
						<label>${__("Client")}</label>
						<div class="client-field"></div>
					</div>
				</div>
			</div>
			<div class="row">
				<div class="col-md-4">
					<div class="form-group">
						<label>${__("Contract Type")}</label>
						<div class="contract-type-field"></div>
					</div>
				</div>
				<div class="col-md-4">
					<div class="form-group">
						<label>${__("Contract Status")}</label>
						<div class="contract-status-field"></div>
					</div>
				</div>
			</div>
			<div class="form-group">
				<label><input type="checkbox" class="create-wbs" checked /> ${__("Create PM WBS Tasks")}</label>
			</div>
			<div class="form-group">
				<label><input type="checkbox" class="submit-baseline" checked /> ${__("Submit Schedule Baseline")}</label>
			</div>
			<div class="form-group">
				<label><input type="checkbox" class="skip-existing" checked /> ${__("Skip already imported projects")}</label>
			</div>
			<div class="form-group">
				<button class="btn btn-secondary btn-sm btn-upload-xer">${__("Upload XER File")}</button>
				<span class="file-label text-muted ml-2"></span>
			</div>
			<div class="preview-area mt-4"></div>
			<div class="mt-3">
				<button class="btn btn-primary btn-import" disabled>${__("Import Selected Projects")}</button>
			</div>
			<div class="results-area mt-4"></div>
		</div>
	`);
	page.main.append($layout);

	const company = frappe.ui.form.make_control({
		parent: $layout.find(".company-field"),
		df: { fieldtype: "Link", options: "Company", fieldname: "company", reqd: 1 },
		render_input: true,
	});
	const branch = frappe.ui.form.make_control({
		parent: $layout.find(".branch-field"),
		df: { fieldtype: "Link", options: "Branch", fieldname: "branch", reqd: 1 },
		render_input: true,
	});
	const client = frappe.ui.form.make_control({
		parent: $layout.find(".client-field"),
		df: { fieldtype: "Link", options: "Customer", fieldname: "client", reqd: 1 },
		render_input: true,
	});
	const contract_type = frappe.ui.form.make_control({
		parent: $layout.find(".contract-type-field"),
		df: {
			fieldtype: "Select",
			fieldname: "contract_type",
			options: "Lump Sum\nUnit Price\nCost Plus\nTurnkey (EPC)",
			default: "Lump Sum",
		},
		render_input: true,
	});
	const contract_status = frappe.ui.form.make_control({
		parent: $layout.find(".contract-status-field"),
		df: {
			fieldtype: "Select",
			fieldname: "contract_status",
			options: "Draft\nActive\nSuspended\nClosed\nCancelled",
			default: "Active",
		},
		render_input: true,
	});

	frappe.db.get_value("Company", { is_group: 0 }, "name").then((r) => {
		if (r?.message?.name) company.set_value(r.message.name);
	});

	$layout.find(".btn-upload-xer").on("click", () => {
		new frappe.ui.FileUploader({
			allow_multiple: false,
			restrictions: { allowed_file_types: [".xer", ".XER", ".txt"] },
			on_success(file) {
				state.file_url = file.file_url;
				$layout.find(".file-label").text(file.file_name || file.file_url);
				preview_file();
			},
		});
	});

	function preview_file() {
		if (!state.file_url) return;
		frappe.call({
			method: "omnexa_construction.primavera_project_import.preview_primavera_xer_import",
			args: { file_url: state.file_url },
			freeze: true,
			callback(r) {
				state.projects = r.message?.projects || [];
				render_preview();
			},
		});
	}

	function render_preview() {
		const $area = $layout.find(".preview-area");
		if (!state.projects.length) {
			$area.html(`<p class="text-muted">${__("No projects found in file.")}</p>`);
			$layout.find(".btn-import").prop("disabled", true);
			return;
		}
		let html = `<h5>${__("Projects in file")}</h5>
			<table class="table table-bordered table-sm">
			<thead><tr>
				<th><input type="checkbox" class="select-all-projects" checked /></th>
				<th>${__("P6 ID")}</th>
				<th>${__("Name")}</th>
				<th>${__("Start")}</th>
				<th>${__("End")}</th>
				<th>${__("Tasks")}</th>
				<th>${__("Value")}</th>
				<th>${__("Status")}</th>
			</tr></thead><tbody>`;
		state.projects.forEach((p, i) => {
			const disabled = p.already_imported ? "disabled" : "";
			const checked = p.already_imported ? "" : "checked";
			html += `<tr>
				<td><input type="checkbox" class="proj-select" data-idx="${i}" ${checked} ${disabled} /></td>
				<td>${frappe.utils.escape_html(p.proj_id)}</td>
				<td>${frappe.utils.escape_html(p.name)}</td>
				<td>${p.start_date || ""}</td>
				<td>${p.end_date || ""}</td>
				<td>${p.task_count || 0}</td>
				<td>${frappe.format(p.contract_value || 0, { fieldtype: "Currency" })}</td>
				<td>${p.already_imported ? __("Already imported") : __("Ready")}</td>
			</tr>`;
		});
		html += "</tbody></table>";
		$area.html(html);
		$layout.find(".btn-import").prop("disabled", false);

		$area.find(".select-all-projects").on("change", function () {
			const on = $(this).is(":checked");
			$area.find(".proj-select:not(:disabled)").prop("checked", on);
		});
	}

	$layout.find(".btn-import").on("click", () => {
		if (!state.file_url) {
			frappe.msgprint(__("Upload an XER file first."));
			return;
		}
		const company_val = company.get_value();
		const branch_val = branch.get_value();
		const client_val = client.get_value();
		if (!company_val || !branch_val || !client_val) {
			frappe.msgprint(__("Company, Branch, and Client are required."));
			return;
		}
		const selected = [];
		$layout.find(".proj-select:checked").each(function () {
			const idx = $(this).data("idx");
			if (state.projects[idx]) selected.push(state.projects[idx].proj_id);
		});
		if (!selected.length) {
			frappe.msgprint(__("Select at least one project."));
			return;
		}
		frappe.call({
			method: "omnexa_construction.primavera_project_import.import_primavera_xer_projects",
			args: {
				file_url: state.file_url,
				company: company_val,
				branch: branch_val,
				client: client_val,
				project_ids: JSON.stringify(selected),
				create_wbs_tasks: $layout.find(".create-wbs").is(":checked") ? 1 : 0,
				submit_baseline: $layout.find(".submit-baseline").is(":checked") ? 1 : 0,
				skip_existing: $layout.find(".skip-existing").is(":checked") ? 1 : 0,
				contract_type: contract_type.get_value(),
				contract_status: contract_status.get_value(),
			},
			freeze: true,
			callback(r) {
				const msg = r.message || {};
				let html = `<h5>${__("Import Results")}</h5><ul>`;
				(msg.results || []).forEach((row) => {
					if (row.status === "success") {
						html += `<li class="text-success">${frappe.utils.escape_html(row.proj_id)} → 
							<a href="/app/project-contract/${row.project_contract}">${row.project_contract}</a>
							(${row.tasks_imported} ${__("tasks")})</li>`;
					} else if (row.status === "skipped") {
						html += `<li class="text-muted">${frappe.utils.escape_html(row.proj_id)} — ${__("skipped")}</li>`;
					} else {
						html += `<li class="text-danger">${frappe.utils.escape_html(row.proj_id)} — ${frappe.utils.escape_html(row.message || "")}</li>`;
					}
				});
				html += `</ul><p><b>${__("Imported")}:</b> ${msg.imported || 0}</p>`;
				$layout.find(".results-area").html(html);
				frappe.show_alert({
					message: __("Imported {0} project(s)", [msg.imported || 0]),
					indicator: "green",
				});
				preview_file();
			},
		});
	});
};

File: omnexa_construction/primavera_project_import.py
# Copyright (c) 2026, Omnexa and contributors
# License: MIT

"""Import Primavera P6 projects from XER into Project Contract + Schedule Baseline."""

from __future__ import annotations

import json

import frappe
from frappe import _
from frappe.utils import add_days, getdate, now_datetime, today

from omnexa_construction.primavera_xer_parser import (
	extract_xer_projects,
	extract_xer_tasks,
	load_xer_content,
	parse_xer_sections,
	project_schedule_bounds,
)


def parse_xer_file(content: str) -> dict[str, list[dict]]:
	return parse_xer_sections(content)


@frappe.whitelist()
def preview_primavera_xer_import(file_url: str) -> dict:
	sections = parse_xer_file(load_xer_content(file_url))
	projects = extract_xer_projects(sections)
	if not projects:
		# Single-project export: derive one project from tasks
		tasks = extract_xer_tasks(sections)
		if not tasks:
			frappe.throw(_("No PROJECT or TASK data found in XER file."), title=_("Primavera Import"))
		start, end = project_schedule_bounds(tasks)
		projects = [
			{
				"proj_id": "XER-IMPORT",
				"name": _("Imported Primavera Schedule"),
				"description": "",
				"start_date": start,
				"end_date": end,
				"contract_value": 0
	}
		]

	preview = []
	for project in projects:
		proj_id = project["proj_id"]
		tasks = extract_xer_tasks(sections, None if proj_id == "XER-IMPORT" else proj_id)
		already = frappe.db.exists("Project Contract", {"p6_project_id": proj_id
	}) if proj_id != "XER-IMPORT" else None
		preview.append(
			{
				**project,
				"start_date": str(project["start_date"]) if project.get("start_date") else None,
				"end_date": str(project["end_date"]) if project.get("end_date") else None,
				"task_count": len(tasks),
				"already_imported": already
	}
		)

	return {
		"projects": preview,
		"total_projects": len(preview),
		"total_tasks": sum(p["task_count"] for p in preview)
	}


@frappe.whitelist()
def import_primavera_xer_projects(
	file_url: str,
	company: str,
	branch: str,
	client: str,
	project_ids: str | None = None,
	create_wbs_tasks: int = 0,
	submit_baseline: int = 1,
	contract_type: str = "Lump Sum",
	contract_status: str = "Active",
	skip_existing: int = 1,
) -> dict:
	if not all([file_url, company, branch, client]):
		frappe.throw(_("Company, Branch, Client, and XER file are required."), title=_("Primavera Import"))

	sections = parse_xer_file(load_xer_content(file_url))
	projects = extract_xer_projects(sections)
	if not projects:
		tasks = extract_xer_tasks(sections)
		if not tasks:
			frappe.throw(_("No importable data in XER file."), title=_("Primavera Import"))
		start, end = project_schedule_bounds(tasks)
		projects = [
			{
				"proj_id": "XER-IMPORT",
				"name": _("Imported Primavera Schedule"),
				"description": "",
				"start_date": start,
				"end_date": end,
				"contract_value": 0
	}
		]

	selected_ids: set[str] | None = None
	if project_ids:
		selected_ids = set(json.loads(project_ids))

	frappe.flags.in_primavera_xer_import = True
	results = []
	try:
		for project in projects:
			proj_id = project["proj_id"]
			if selected_ids is not None and proj_id not in selected_ids:
				continue

			if skip_existing and proj_id != "XER-IMPORT":
				existing = frappe.db.get_value("Project Contract", {"p6_project_id": proj_id
	}, "name")
				if existing:
					results.append(
						{
							"proj_id": proj_id,
							"status": "skipped",
							"project_contract": existing,
							"message": _("Already imported")
	}
					)
					continue

			try:
				result = _import_single_project(
					project=project,
					sections=sections,
					company=company,
					branch=branch,
					client=client,
					create_wbs_tasks=bool(int(create_wbs_tasks)),
					submit_baseline=bool(int(submit_baseline)),
					contract_type=contract_type,
					contract_status=contract_status,
				)
				results.append(result)
				frappe.db.commit()
			except Exception as exc:
				frappe.db.rollback()
				frappe.log_error(title=_("Primavera XER import failed"), message=frappe.get_traceback())
				results.append(
					{
						"proj_id": proj_id,
						"status": "failed",
						"message": str(exc)
	}
				)
	finally:
		frappe.flags.in_primavera_xer_import = False

	imported = sum(1 for r in results if r.get("status") == "success")
	return {
		"imported": imported,
		"results": results
	}


def _import_single_project(
	project: dict,
	sections: dict,
	company: str,
	branch: str,
	client: str,
	create_wbs_tasks: bool,
	submit_baseline: bool,
	contract_type: str,
	contract_status: str,
) -> dict:
	proj_id = project["proj_id"]
	tasks = extract_xer_tasks(sections, None if proj_id == "XER-IMPORT" else proj_id)
	start, end = project_schedule_bounds(tasks, project.get("start_date"), project.get("end_date"))

	contract = frappe.get_doc(
		{
			"doctype": "Project Contract",
			"contract_title": project["name"],
			"contract_type": contract_type or "Lump Sum",
			"client": client,
			"company": company,
			"branch": branch,
			"status": contract_status or "Active",
			"contract_value": project.get("contract_value") or 0,
			"planned_start": start,
			"planned_completion": end,
			"site_location": (project.get("description") or "")[:140] or None
	}
	)
	if frappe.get_meta("Project Contract").has_field("p6_project_id") and proj_id != "XER-IMPORT":
		contract.p6_project_id = proj_id
		if frappe.get_meta("Project Contract").has_field("p6_sync_status"):
			contract.p6_sync_status = "Synced"
		if frappe.get_meta("Project Contract").has_field("p6_last_sync"):
			contract.p6_last_sync = now_datetime()
	contract.insert(ignore_permissions=True)

	baseline = frappe.get_doc(
		{
			"doctype": "Construction Schedule Baseline",
			"project_contract": contract.name,
			"baseline_date": today(),
			"baseline_name": _("Primavera Import — {0}").format(project["name"])[:140],
			"planned_start": start,
			"planned_completion": end,
			"is_active": 1,
			"company": company,
			"branch": branch,
			"notes": _("Imported from Primavera XER on {0}").format(today())
	}
	)
	for task in tasks:
		row = {
			"task_name": task["task_name"],
			"start_date": task["start_date"],
			"end_date": task["end_date"],
			"duration_days": task["duration_days"],
			"is_milestone": task.get("is_milestone") or 0,
			"cost_code": task.get("cost_code")
	}
		meta = frappe.get_meta("Construction Schedule Baseline Task")
		if meta.has_field("predecessor_task") and task.get("predecessor_task"):
			row["predecessor_task"] = task["predecessor_task"]
		baseline.append("tasks", row)

	baseline.insert(ignore_permissions=True)
	if submit_baseline:
		baseline.submit()

	wbs_count = 0
	if create_wbs_tasks and tasks:
		wbs_count = _create_pm_wbs_tasks(contract.name, company, branch, tasks)

	return {
		"proj_id": proj_id,
		"status": "success",
		"project_contract": contract.name,
		"baseline": baseline.name,
		"tasks_imported": len(tasks),
		"wbs_tasks_created": wbs_count
	}


def _create_pm_wbs_tasks(project_contract: str, company: str, branch: str, tasks: list[dict]) -> int:
	meta = frappe.get_meta("PM WBS Task")
	count = 0
	for idx, task in enumerate(tasks, start=1):
		doc = {
			"doctype": "PM WBS Task",
			"project": project_contract,
			"task_name": task["task_name"],
			"planned_start": task["start_date"],
			"planned_end": task["end_date"],
			"sequence_no": idx,
			"status": "Planned",
			"company": company,
			"branch": branch
	}
		if meta.has_field("p6_activity_id") and task.get("task_id"):
			doc["p6_activity_id"] = task["task_id"]
		if meta.has_field("p6_sync_status"):
			doc["p6_sync_status"] = "Synced"
		frappe.get_doc(doc).insert(ignore_permissions=True)
		count += 1
	return count
